mail/spring-cloud-starter-stream-source-mail/src/main/java/org/springframework/cloud/stream/app/mail/source/MailSourceConfiguration.java

```diff
@@ -0,0 +1,160 @@
+import org.springframework.integration.dsl.mail.Mail;
+import org.springframework.integration.dsl.mail.MailInboundChannelAdapterSpec;
+import org.springframework.integration.dsl.support.Consumer;
+import org.springframework.integration.mail.transformer.MailToStringTransformer;
+import org.springframework.integration.scheduling.PollerMetadata;
+
+/**
+ * A source module that listens for mail and emits the content as a message payload.
+ *
+ * @author Amol
+ */
+@EnableBinding(Source.class)
+@EnableConfigurationProperties({ MailSourceProperties.class })
+@Import({ TriggerConfiguration.class })
+public class MailSourceConfiguration {
+
+	@Autowired
+	@Qualifier("defaultPoller")
+	PollerMetadata defaultPoller;
+
+	@Autowired
+	@Bindings(MailSourceConfiguration.class)
+	Source source;
+
+	@Autowired
+	MailSourceProperties properties;
+
+	@Bean
+	public IntegrationFlow mailInboundFlow() {
+
+		IntegrationFlowBuilder flowBuilder;
+
+		flowBuilder = getFlowBuilder();
+
+		return flowBuilder.transform(new MailToStringTransformer())
+				.channel(source.output()).get();
+	}
+
+	/**
+	 * Method to build Integration Flow for Mail. Suppress Warnings for
+	 * MailInboundChannelAdapterSpec.
+	 * @return Integration Flow object for Mail Source
+	 */
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	private IntegrationFlowBuilder getFlowBuilder() {
+		Properties javaMailProperties = new Properties();
+		String mailURL = properties.getProtocol() + "://" + properties.getUsername() + ":"
+				+ properties.getPassword() + "@" + properties.getHost() + ":"
+				+ properties.getPort() + "/" + properties.getFolder();
+
+		javaMailProperties.putAll(properties.parsePropertiesString());
+
+		IntegrationFlowBuilder flowBuilder;
+		if (properties.isIdleImap()) {
+			flowBuilder = getIdleImapflow(mailURL);
+		}
+		else {
+
+			MailInboundChannelAdapterSpec adapterSpec;
+			switch (properties.getProtocol().toUpperCase()) {
+			case "IMAP":
+			case "IMAPS":
+				adapterSpec = getImapFlowBuilder(mailURL);
+				break;
+			case "POP3":
+				adapterSpec = getPop3FlowBuilder(mailURL);
+				break;
+			default:
+				adapterSpec = null;
+				break;
+			}
+
+			flowBuilder = IntegrationFlows.from(
+					adapterSpec.javaMailProperties(javaMailProperties)
+							.shouldDeleteMessages(properties.isDelete()),
+					new Consumer<SourcePollingChannelAdapterSpec>() {
+
+						@Override
+						public void accept(
+								SourcePollingChannelAdapterSpec sourcePollingChannelAdapterSpec) {
+							sourcePollingChannelAdapterSpec.poller(defaultPoller);
+						}
+					});
+
+		}
+		return flowBuilder;
+	}
+
+	/**
+	 * Method to build Integration flow for IMAP Idle configuration.
+	 * @param mailURL Mail source URL.
+	 * @return Integration Flow object IMAP IDLE.
+	 */
+	private IntegrationFlowBuilder getIdleImapflow(String mailURL) {
+		IntegrationFlowBuilder flowBuilder = null;
+		if ("imap".equalsIgnoreCase(properties.getProtocol())
+				|| "imaps".equalsIgnoreCase(properties.getProtocol())) {
```

Comment:
I think this `if` does not make sense if we use `idleImap = true`.
The "wrong protocol" is a responsibility of the target `Mail.imapIdleAdapter()` already.